pkg/planner/core/operator/logicalop/logical_aggregation.go

```diff
@@ -632,49 +632,105 @@ func (la *LogicalAggregation) splitCondForAggregation(predicates []expression.Ex
+				}
+				ret = append(ret, subRet1...)
+			}
 		}
 	}
 	return condsToPush, ret
 }
 
-// pushDownPredicatesForAggregation split a condition to two parts, can be pushed-down or can not be pushed-down below aggregation.
-func (la *LogicalAggregation) pushDownPredicatesForAggregation(cond expression.Expression, groupByColumns *expression.Schema, exprsOriginal []expression.Expression) (condsToPush, ret []expression.Expression) {
-	switch cond.(type) {
+// getAggFuncsColsForFirstRow gets the columns that are used by first_row agg functions.
+func (la *LogicalAggregation) getAggFuncsColsForFirstRow() (aggFuncsCols []*expression.Column) {
+	aggFuncsCols = make([]*expression.Column, 0, len(la.AggFuncs))
+	for idx, col := range la.Schema().Columns {
+		aggFunc := la.AggFuncs[idx]
+		if aggFunc.Name == ast.AggFuncFirstRow {
+			cols := expression.ExtractColumns(aggFunc.Args[0])
+			if len(cols) == 1 {
+				aggFuncsCols = append(aggFuncsCols, col)
+			}
+		}
+	}
+	return aggFuncsCols
+}
+
+// pushDownPredicates split a condition to two parts, can be pushed-down or can not be pushed-down below aggregation.
+func (*LogicalAggregation) pushDownPredicates(cond expression.Expression, groupByColumns *expression.Schema, exprsOriginal []expression.Expression, pushDownFunc pushDownPredicatesFunc) (condsToPush, ret []expression.Expression) {
+	switch c := cond.(type) {
 	case *expression.Constant:
 		condsToPush = append(condsToPush, cond)
 		// Consider SQL list "select sum(b) from t group by a having 1=0". "1=0" is a constant predicate which should be
 		// retained and pushed down at the same time. Because we will get a wrong query result that contains one column
 		// with value 0 rather than an empty query result.
 		ret = append(ret, cond)
 	case *expression.ScalarFunction:
-		extractedCols := expression.ExtractColumns(cond)
-		ok := true
-		for _, col := range extractedCols {
-			if !groupByColumns.Contains(col) {
-				ok = false
-				break
-			}
-		}
-		if ok {
-			newFunc := expression.ColumnSubstitute(la.SCtx().GetExprCtx(), cond, la.Schema(), exprsOriginal)
-			condsToPush = append(condsToPush, newFunc)
-		} else {
-			ret = append(ret, cond)
-		}
+		return pushDownFunc(c, groupByColumns, exprsOriginal)
 	default:
 		ret = append(ret, cond)
 	}
 	return condsToPush, ret
 }
 
+type pushDownPredicatesFunc func(cond *expression.ScalarFunction, cols *expression.Schema, exprsOriginal []expression.Expression) (_, _ []expression.Expression)
+
+func (la *LogicalAggregation) pushDownPredicatesByGroupby(cond *expression.ScalarFunction, groupByColumns *expression.Schema, exprsOriginal []expression.Expression) (condsToPush, ret []expression.Expression) {
+	extractedCols := expression.ExtractColumns(cond)
+	ok := true
+	for _, col := range extractedCols {
+		if !groupByColumns.Contains(col) {
+			ok = false
+			break
+		}
+	}
+	if ok {
+		newFunc := expression.ColumnSubstitute(la.SCtx().GetExprCtx(), cond, la.Schema(), exprsOriginal)
```

Comment:
for previous code:
if con's cols is all from groupByColumns,  that's means there is nothing to be substituted from la.Schema() with the agg func's original arg[0].

Unless, the group by column is also kept as first_row() func in agg itself, so substituttion: becomes: (gbyCol+ x) with schema col is gbyCol <= first(gbyCol) arg[0] = gbyCol, so we got (gbyCol+ x), currently i don't think out of any other case, maybe the old logic ColumnSubstitute can be eliminated here?

it's aspirational requirement, it's ok /cc @hawkingrei @ghoshh